### ISSUE TITLE
[Consolidated] Segfaults

### ISSUE THREAD

**Issue by wintered** [Z3Prover/z3#6176]
[81cb575](https://github.com/Z3Prover/z3/commit/81cb575c22b01ba376d2ec6d607481686011d9ee)
```
$z3release bug.smt2                                                             
[1]    3360378 segmentation fault  z3release bug.smt2                           
$cat bug.smt2                                                                   
(declare-fun a () String)                                                       
(set-option :proof true)                                                        
(assert (and (not (not (= 0 0))) (not (= 0 0))))                                
(check-sat-using qfufbv_ackr) 
```

**Comment by wintered** [comment]
[81cb575](https://github.com/Z3Prover/z3/commit/81cb575c22b01ba376d2ec6d607481686011d9ee)
```
$z3release bug.smt2                                                             
[1]    802442 segmentation fault  z3release bug.smt2                            
$cat bug.smt2                                                                   
(declare-fun value2 () String)                                                  
(set-option :proof true)                                                        
(assert (and (= 0 0) (not (= 0 0))))                                            
(check-sat-using qffp) 
```

**Comment by wintered** [comment]
[81cb575](https://github.com/Z3Prover/z3/commit/81cb575c22b01ba376d2ec6d607481686011d9ee)
```
$z3release horn.smt2                                                            
[1]    4083342 segmentation fault  z3release horn.smt2                          
$cat horn.smt2                                                                  
(assert (forall ((a Bool) (b Bool) (d (_ BitVec 1)) (e (_ BitVec 1)) (f (_ BitVec 1)) (g Bool))
(= (= f (ite a (_ bv0 1) (ite true (_ bv0 1) (ite b e e)))) g)))                
(check-sat-using horn) 
```

**Comment by NikolajBjorner** [comment]
I think the reality check here is "who cares" about these bugs?
You pass a non-Boolean argument to interpolation, which isn't even advertised functionality?
I fixed this one, but then who benefits vs. the overhead fixing?

**Comment by levnach** [comment]
z3 from the Nightly build crashes on the following input file(s).

**z3 Nightly build:** tag `Nightly`, commit [`241c6211d`](https://github.com/Z3Prover/z3/commit/241c6211d6e0da7dd4ef8904314aa3db0775db30) (`Z3 version 4.17.0`)

**Crashing input(s):**
- [`inputs/issues/iss-6176/horn.smt2`](https://github.com/Z3Prover/bench/blob/main/inputs/issues/iss-6176/horn.smt2)
- [`inputs/issues/iss-6176/horn-2.smt2`](https://github.com/Z3Prover/bench/blob/main/inputs/issues/iss-6176/horn-2.smt2)

z3 crashes when run on the file(s) above.

The crash is observed on Ubuntu 24.04.4 LTS (x86_64).

**Comment by NikolajBjorner** [comment]
## Verification (2026): all reported segfaults are fixed

Re-tested every repro from this thread on current master (**Z3 4.17.0**, build `d99b6e7b0`), on both the release and a debug build (assertions enabled). None crash:

| Repro | Config | Result now |
|-------|--------|-----------|
| `horn.smt2` | `(check-sat-using horn)` on forall/bitvec | `unsat`, no crash |
| `horn-2.smt2` | duplicate of `horn.smt2` | `unsat`, no crash |
| `bug.smt2` (body) | `(check-sat-using qfufbv_ackr)` + `:proof true` | `unsat`, no crash |
| `bug.smt2` (comment) | `(check-sat-using qffp)` + `:proof true` | `unsat`, no crash |
| interpolant repro | `(get-interpolant a true)`, non-Boolean arg | graceful error, no crash |

### The horn crash

The only case still reported crashing (comment above from 2026-06-07 at commit `241c6211d`) was **fixed by commit `14e1dc989` — "Fix segfault in horn on formulas with unused quantified variables (#10091)"**. That commit is a descendant of `241c6211d`, so the report simply predates the fix.

Root cause: in `horn.smt2` the quantified body `(= (= f (ite a (_ bv0 1) (ite true (_ bv0 1) (ite b e e)))) g)` simplifies (`ite true X Y -> X`, `ite c z z -> z`) so that variables `d`, `b`, `e` become unused. The Horn query builder's single-pass gap-elimination for unused quantified variables then left a null sort behind, which got dereferenced. The fix iterates gap-elimination to a fixpoint (the rewriter's ITE collapsing can create new gaps).

The `qfufbv_ackr`/`qffp` + `:proof true` cases and the interpolation case were fixed earlier.

Closing as resolved — all listed segfaults now terminate cleanly on master.
